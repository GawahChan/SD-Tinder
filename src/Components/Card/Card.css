.card {
    position: absolute;
    border-radius: 10px;
    box-shadow: 2px 2px 20px #202020;

    background-position: top;
    background-size: cover;
    overflow: hidden;

    display: flex;
    flex-direction: column;
    justify-content: flex-end; 

    width: 320px;
    height: 500px;

    transition: box-shadow 0.5s
}

.card img {
    vertical-align: middle;
}

.card h1 {
    margin: 0;
    font-weight: 600;
    font-size: 1.5rem;
}

.card p {
    margin: 0;
}

/*Display Action Overlay CSS */
.display-action {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 50%;
}

.display-action h1 {
    font-size: 2.5rem;
    font-weight: 500;
}

.display-action.liked{
    color: #4bb12c;
}

.display-action.superliked{
    color: #09cfcf
}

.display-action.disliked{
    color: #ff0000
}

/*Card contents CSS*/
.card-content {
    padding: 1rem;
    background: linear-gradient(rgba(0,0,0,0), rgba(0,0,0,0.7),rgba(0,0,0,0.9), rgba(0,0,0,1));
    color: #DDDDDD
}

.card-subtitle {
    display: flex;
    justify-content: space-between;
}

.card-subtitle button {
    background: none;
    padding: 10px;
    border: none;
    outline: 0;
    cursor: pointer;
}
.card-actions {
    display: flex;
    justify-content: center;
}

.card-actions button {
    background: none;
    border: 2px solid #DDDDDD;
    border-radius: 50%;
    cursor: pointer;
    margin: 5px;
    height: 3rem;
    width: 3rem;
    outline: 0;
}

/*Event handler CSS */
.card.liked {
    box-shadow: 5px 5px 100px 15px #4bb12c;
}

.card.disliked {
    box-shadow: 5px 5px 100px 15px #ff0000
}

.card.superliked {
    box-shadow: 5px 5px 100px 15px #09cfcf
}

/*Animations */
.card button:hover {
    animation: pulse 3s infinite;
}

@keyframes pulse {
    0% { transform: scale(1)}
    50% { transform: scale(1.1)}
    100% {transform: scale(1)}
}