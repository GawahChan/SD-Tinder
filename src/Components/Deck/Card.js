import React, { useState } from "react";
import PropTypes from "prop-types";
import { CardContainer } from "./style";
import CardDisplay from "./CardDisplay";

import {
  INITIAL_CARD_POSITION,
  REMOVE_LIKED_ANIMATION,
  REMOVE_DISLIKED_ANIMATION,
  REMOVE_SUPERLIKED_ANIMATION,
  DISPLAY_SUPERLIKED_BOUNDARY,
  DISPLAY_LIKED_BOUNDARY,
  DISPLAY_DISLIKED_BOUNDARY,
  REMOVE_SUPERLIKED_BOUNDARY,
  REMOVE_LIKED_BOUNDARY,
  REMOVE_DISLIKED_BOUNDARY
} from "../../Constants";

const Card = ({ stuntDouble, handleAction }) => {
  const [action, setAction] = useState("");
  const [removeCard, setRemoveCard] = useState(INITIAL_CARD_POSITION);

  const animateAction = (id, action) => {
    let removeCard =
      action === "liked"
        ? REMOVE_LIKED_ANIMATION
        : action === "disliked"
        ? REMOVE_DISLIKED_ANIMATION
        : action === "superliked"
        ? REMOVE_SUPERLIKED_ANIMATION
        : INITIAL_CARD_POSITION;

    setAction(action);
    setRemoveCard(removeCard);

    handleAction(id, action);
  };

  const handleDrag = info => {
    let xPosition = info.point.x;
    let yPosition = info.point.y;

    let action =
      xPosition > DISPLAY_LIKED_BOUNDARY
        ? "liked"
        : xPosition < DISPLAY_DISLIKED_BOUNDARY
        ? "disliked"
        : yPosition < DISPLAY_SUPERLIKED_BOUNDARY
        ? "superliked"
        : "";

    setAction(action);
  };

  const handleDragEnd = (id, info) => {
    let xPosition = info.point.x;
    let yPosition = info.point.y;

    return xPosition > REMOVE_LIKED_BOUNDARY
      ? animateAction(id, "liked")
      : xPosition < REMOVE_DISLIKED_BOUNDARY
      ? animateAction(id, "disliked")
      : yPosition < REMOVE_SUPERLIKED_BOUNDARY
      ? animateAction(id, "superliked")
      : setAction("");
  };

  return (
    <CardContainer
      drag
      dragElastic={0.8}
      dragConstraints={{ top: 0, right: 0, bottom: 0, left: 0 }}
      onDrag={(e, info) => handleDrag(info)}
      onDragEnd={(e, info) => handleDragEnd(stuntDouble.id, info)}
      initial={{ opacity: 0 }}
      animate={{ opacity: 1 }}
      exit={removeCard}
      actionColour={action}
    >
      <CardDisplay
        stuntDouble={stuntDouble}
        animateAction={animateAction}
        action={action}
      />
    </CardContainer>
  );
};

Card.propTypes = {
  stuntDouble: PropTypes.object,
  handleAction: PropTypes.func
};

export default Card;
