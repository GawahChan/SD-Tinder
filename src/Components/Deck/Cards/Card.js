import React, { Component } from 'react';
import { CardContainer } from './style';
import CardDisplay from './CardDisplay';

class Card extends Component {
    constructor(props) {
        super(props);
        this.state = {
            action: '',
            removeCard: false
        }
    }

    animateAction = (id, action) => {
        let handleCardAnimation =
            action === 'liked' ? { x: 400, opacity: 0 } :
                action === 'disliked' ? { x: -400, opacity: 0 } :
                    action === 'superliked' ? { y: -400, opacity: 0 } :
                        { x: 0, y: 0 }

        this.setState({ action, removeCard: handleCardAnimation })

        this.props.handleAction(id, action);
    }

    handleDrag = (info) => {
        let xPosition = info.point.x;
        let yPosition = info.point.y;

        let action =
            xPosition > 30 ? 'liked' :
                xPosition < -30 ? 'disliked' :
                    yPosition < -30 ? 'superliked' : '';

        this.setState({ action });
    }

    handleDragEnd = (id, info) => {
        let xPosition = info.point.x;
        let yPosition = info.point.y;

        return (
            xPosition > 80 ? this.animateAction(id, 'liked') :
                xPosition < -80 ? this.animateAction(id, 'disliked') :
                    yPosition < -80 ? this.animateAction(id, 'superliked') :
                        this.setState({ action: '' })
        );
    }
    render() {
        return (
            <CardContainer
                drag
                dragElastic={0.8}
                dragConstraints={{ top: 0, right: 0, bottom: 0, left: 0 }}
                onDrag={(e, info) => this.handleDrag(info)}
                onDragEnd={(e, info) => this.handleDragEnd(this.props.stuntDouble.id, info)}
                animate={this.state.removeCard}
                actionColour={this.state.action}
            >
                <CardDisplay
                    stuntDouble={this.props.stuntDouble}
                    animateAction={this.animateAction}
                    action={this.state.action}
                />
            </CardContainer >
        );
    }
}

export default Card;