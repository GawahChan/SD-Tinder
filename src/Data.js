const Data = [
    {id: 0, name: 'Mila Kunis', age: '30', location: 'LA', bio: 'Actress' }
];

export default Data;