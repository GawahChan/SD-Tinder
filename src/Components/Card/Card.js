import React from 'react';
import Dislike from '../../Common/Icons/Dislike.svg';
import Superlike from '../../Common/Icons/SuperLike.svg';
import Like from '../../Common/Icons/Like.svg';

import './Card.css';

function Card(props) {
    return (
        <div className='card'>
            <div className='card-img' >
                <img src={require(`../../Common/Profile/${props.name}.jpg`)} alt='Profile' />
            </div>
            <div className='card-bio'>
                <h1>{props.name}, {props.age}</h1>
                <p>{props.location}</p>
                <p>{props.bio}</p>
            </div>
            <div className='card-actions'>
                <button><img src={Dislike} alt='Dislike' /></button>
                <button><img src={Superlike} alt='Superlike' /></button>
                <button onClick={() => props.liked(props.id)}><img src={Like} alt='Like' /></button>
            </div>
        </div>
    );
}

export default Card;