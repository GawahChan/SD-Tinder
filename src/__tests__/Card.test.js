import React from 'react';
import { shallow } from 'enzyme'
import Card from '../Components/Card/Card';
import Data from '../Data';

describe('Card Component', () => {
    const mockHandleAction = jest.fn()
    const props = Data[0];
   
    const card = shallow(<Card {...props} handleAction = {mockHandleAction} />);

    it('renders correctly', () => {
        expect(card).toMatchSnapshot();
    });

    it('initialises the `display bio` state as false', () => {
        expect(card.state('displayBio')).toBe(false);
    });

    describe('user clicks an action button', () => {
        afterEach(() => {
            mockHandleAction.mockClear();
        });

        it('disliked button calls event handler', () => {
            card.find('.disliked').simulate('Click', { currentTarget: {className: 'disliked'}})

            expect(mockHandleAction).toHaveBeenCalledTimes(1);
        })

        it('superliked button calls event handler', () => {
            card.find('.superliked').simulate('Click', { currentTarget: {className: 'superliked'}})

            expect(mockHandleAction).toHaveBeenCalledTimes(1);
        });

        it('liked button calls event handler', () => {
            card.find('.liked').simulate('Click', { currentTarget: {className: 'liked'}})

            expect(mockHandleAction).toHaveBeenCalledTimes(1);
        });
    });

    describe('toggleBio event handler', () => {
        beforeEach(() => {
            card.instance().toggleBio();
        })

        afterEach(() => {
            card.setState({displayBio: false});
        })
        it('toggles the `displayBio` state', () => {
            expect(card.state('displayBio')).toBe(true);
        });
    });
});