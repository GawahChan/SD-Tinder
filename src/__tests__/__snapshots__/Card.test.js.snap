// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component renders correctly 1`] = `
<ForwardRef(MotionComponent)
  animate={false}
  className="card "
  drag={true}
  dragConstraints={
    Object {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    }
  }
  dragElastic={0.5}
  key="0"
  onDrag={[Function]}
  onDragEnd={[Function]}
  style={
    Object {
      "backgroundImage": "url(Alicia-Vela-Bailey.jpg)",
    }
  }
  whileTap={
    Object {
      "scale": 0.9,
    }
  }
>
  <div
    className="display-action "
  >
    <h1 />
  </div>
  <div
    className="card-content"
  >
    <h1>
      Alicia Vela-Bailey
      , 
      28
    </h1>
    <div
      className="card-subtitle"
    >
      <p>
        LA, USA
      </p>
      <button
        onClick={[Function]}
      >
        <img
          alt="Info"
          src="info.svg"
        />
      </button>
    </div>
    <div
      className="card-actions"
    >
      <button
        className="disliked"
        onClick={[Function]}
      >
        <img
          alt="Dislike"
          src="Dislike.svg"
        />
      </button>
      <button
        className="superliked"
        onClick={[Function]}
      >
        <img
          alt="Superlike"
          src="SuperLike.svg"
        />
      </button>
      <button
        className="liked"
        onClick={[Function]}
      >
        <img
          alt="Like"
          src="Like.svg"
        />
      </button>
    </div>
  </div>
</ForwardRef(MotionComponent)>
`;
